### PR TITLE
Ryna - LB112.ag_prodchange_R_C_Y_GLU

### PR DESCRIPTION
**Pull request by rynacui**

### CODE REVIEW

**Comment by rynacui** [comment]
@bpbond I have two errors that says the `GCAM_region_ID` in ODS are numeric, associated with the outputs `L112.ag_YieldRate_R_C_Y_GLU.csv` and `L112.bio_YieldRate_R_Y_GLU.csv` respectively.

This is very similar to what I got in the last chunk where `year` shown as numeric in the ODS and failed the test for me but not for you. I'm wondering if this is happening again, although I've updated `readr` to v1.1. 

Could you please try it on your machine? Thanks!

File: R/zchunk_LB112.ag_prodchange_R_C_Y_GLU.R [hunk 1]
```diff
@@ -40,62 +41,292 @@ module_aglu_LB112.ag_prodchange_R_C_Y_GLU_DISABLED <- function(command, ...) {
...
+      # Replace crop IDs with names
+      left_join_error_no_match(crop_id, by = c("crop_ID" = "CROSIT_cropID")) %>%
+      # NOTE: Only use total production and harvested area from CROSIT (not separating rainfed and irrigated).
+      select(CROSIT_ctry, CROSIT_crop, year, HA_kha, Yield_kgHa, Prod_kt,
+             -HA_kha_rainfed, -Yield_kgHa_rainfed, -Prod_kt_rainfed, -HA_kha_irrigated, -Yield_kgHa_irrigated, -Prod_kt_irrigated) %>%
+      # Convert the needed numeric variables from integer to double for calculation
+      mutate(HA_kha = as.double(HA_kha),
+             Yield_kgHa = as.double(Yield_kgHa),
+             Prod_kt = as.double(Prod_kt)) %>%
+      # Some regions have 0 production with positive harvested area and a reported yield.
```
Comment (rynacui) [inline R/zchunk_LB112.ag_prodchange_R_C_Y_GLU.R]: Starting here in this pipe, there are several adjustments made of production and harvest area data. However, it seems like these two variables are not used for anything else in this chunk, and removing these lines does not affect any outputs.

**Comment by bpbond** [comment]
Hi @rynacui - this is a deceptive error message, and I'm going to push a better one to this PR. It looks like the dimensions of the data frames aren't identical, and that's the fundamental cause of the problem.

```R
> x[["L112.ag_YieldRate_R_C_Y_GLU"]] %>% spread(year, value) %>% arrange(GLU, GCAM_commodity) %>% dim
[1] 3207   22
dim(readr::read_csv("tests/testthat/comparison_data/aglu/L112.ag_YieldRate_R_C_Y_GLU.csv", comment = "#"))
[1] 3596   22
>

File: inst/extdata/aglu/A_defaultYieldRate.csv [hunk 2]
```diff
@@ -0,0 +1,18 @@
+# File: A_defaultYieldRate.csv,,,,,,,,,,,,,,,
+# Title: Reference improvement rates of each crop's maximum yield trajectory,,,,,,,,,,,,,,,
+# Unit: NA,,,,,,,,,,,,,,,
+# Description: Reference improvement rates of each crop's maximum yield trajectory,,,,,,,,,,,,,,,
+GCAM_commodity,2005,2010,2020,2035,2050,2065,2080,2095,2110,2125,2140,2155,2170,2185,2200
```
Comment (bpbond) [inline inst/extdata/aglu/A_defaultYieldRate.csv]: Source of these data?

File: R/zchunk_LB112.ag_prodchange_R_C_Y_GLU.R [hunk 3]
```diff
@@ -40,62 +41,292 @@ module_aglu_LB112.ag_prodchange_R_C_Y_GLU_DISABLED <- function(command, ...) {
...
+      L112.agBio_YieldRatio
+
+    # Calculate the yield change rates for all specified ag productivity years
+    # First prepare a yield rate table for rows binding in the loop
+    L112.agBio_YieldRatio %>%
+      mutate(YieldRate = NA) %>%
+      select(-YieldRatio) -> L112.agBio_YieldRate
+
+    # For each model timestep between 2010 to 2050, calculate the annual yield change rate based on the yield ratios
+    for(i in 2:length(SPEC_AG_PROD_YEARS)) {
```
Comment (bpbond) [inline R/zchunk_LB112.ag_prodchange_R_C_Y_GLU.R]: Maybe add `assert_that(length(SPEC_AG_PROD_YEARS) > 1)` above this line.

File: R/zchunk_LB112.ag_prodchange_R_C_Y_GLU.R [hunk 4]
```diff
@@ -40,62 +41,292 @@ module_aglu_LB112.ag_prodchange_R_C_Y_GLU_DISABLED <- function(command, ...) {
...
+    # Fill out missing years / commodity / region, using the default ag productivity change assumptions
+    # Match these annual improvement rates into a table of existing crop yields, get all GCAM region x commodity x GLU combinations in the production table
+    L103.ag_Prod_Mt_R_C_Y_GLU %>%
+      select(GCAM_region_ID, GCAM_commodity, GLU) %>%
+      unique() %>%
+      # Join the yield change rates for all specified ag productivity years from 2010 to 2050 (this creates NAs, use left_join instead of left_join_error_no_match)
+      left_join(L112.agBio_YieldRate, by = c("GCAM_region_ID", "GCAM_commodity", "GLU")) %>%
+      # Join the default ag producivity improvement assumptions for all future years
+      full_join(L112.defaultYieldRate, by = c("GCAM_commodity", "year")) %>%
+      # Keep futuer years (2015-2100) and 2010
```
Comment (bpbond) [inline R/zchunk_LB112.ag_prodchange_R_C_Y_GLU.R]: "future" (also below)

**Comment by rynacui** [comment]
@kvcalvin Please hold your review till I clean the test. Sorry it's not ready yet!

**Comment by rynacui** [comment]
@bpbond I've revised the code so you may have to look at it again starting at line 208. I misunderstood some parts in the ODS. 
But I'm still getting the error `Incompatible type for column GCAM_region_ID: x numeric, y matrix`. I've checked the outputs look fine with `GCAM_region_ID` as integer. Do you mind running it? Thanks a lot!

**Comment by bpbond** [comment]
We definitely have some bug in the test code that's producing that error, and I will try to find it. But the data frames are also definitely different (see below). I think?

```R
> summary(readr::read_csv("tests/testthat/comparison_data/aglu/L112.ag_YieldRate_R_C_Y_GLU.csv", comment="#") %>% gather(year, value, -GCAM_region_ID, -GCAM_commodity, -GLU))
 GCAM_region_ID  GCAM_commodity         GLU                year               value          
 Min.   : 1.00   Length:68324       Length:68324       Length:68324       Min.   :-0.493846  
 1st Qu.: 9.00   Class :character   Class :character   Class :character   1st Qu.: 0.001000  
 Median :14.00   Mode  :character   Mode  :character   Mode  :character   Median : 0.002000  
 Mean   :15.36                                                            Mean   : 0.050194  
 3rd Qu.:23.00                                                            3rd Qu.: 0.005453  
 Max.   :32.00                                                            Max.   : 1.000000  

>summary(readr::read_csv("outputs/L112.ag_YieldRate_R_C_Y_GLU.csv", skip=1, comment="#"))
 GCAM_region_ID  GCAM_commodity         GLU                 year          value        
 Min.   : 1.00   Length:68324       Length:68324       Min.   :2010   Min.   :-0.9888  
 1st Qu.: 9.00   Class :character   Class :character   1st Qu.:2030   1st Qu.: 0.0010  
 Median :14.00   Mode  :character   Mode  :character   Median :2055   Median : 0.0025  
 Mean   :15.36                                         Mean   :2055   Mean   : 0.1142  
 3rd Qu.:23.00                                         3rd Qu.:2080   3rd Qu.: 0.1257  
 Max.   :32.00                                         Max.   :2100   Max.   : 1.7118  
```

**Comment by rynacui** [comment]
"Apparently", I need a parenthesis in the mathematical formula. 
Sorry about that. Since the test only gives me the numeric-matrix error, had no idea what was going on...

**Comment by bpbond** [comment]
Ok! I will try and fix the test error.


On Apr 28, 2017, at 3:44 PM, rynacui <notifications@github.com<mailto:notifications@github.com>> wrote:


"Apparently", I need a parenthesis in the mathematical formula.
Sorry about that. Since the test only gives me the numeric-matrix error, had no idea what was going on...

**Comment by bpbond** [comment]
@rynacui - I think I have fixed the test code bug in commit 66d6f33.

**Comment by rynacui** [comment]
Clean test now. This PR is ready for review @kvcalvin @bpbond

**Comment by bpbond** [comment]
Awesome! 👍

**Comment by rynacui** [comment]
Hi @bpbond, why is this PR closed but not merged?

**Comment by bpbond** [comment]
Oh crap. Because I screwed up. Sorry!

**Comment by bpbond** [comment]
This is holding on your review/approval @kvcalvin

**Pull request merged**